### PR TITLE
store/tikv: rewrite the 2PC to reduce memory usage

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Problem Summary:

The current transaction's commit implementation collects all the mutations in the `initKeysAndMutations` function before 2PC. In a large transaction, the memory usage of the mutations would be the dominant part of the whole memory occupation.

### What is changed and how it works?

I propose to abstract the mutations to a stream, and during the 2PC processing, only part of the mutations are visited, so as to cut down the memory usage.

I've verified in https://github.com/pingcap/tidb/pull/15105, the result is amazing!
Work together with disk-persistent mutations, we can support very large transactions with small memory. 

What's Changed:

The mutations and locks are abstract to an iterator using merge sort,
`takeBatchMutationFromIterator` takes a batch of the mutations and send it to the worker.

The optimistic transaction uses the new implementation.
The pessimistic transaction is not changed yet. For one reason, this commit is complex enough, I don't want to handle it here; for another, the primary key of pessimistic transaction may not be the first one, makes it differ from the optimistic transaction.

How it Works:

The task channel is unbuffered, so the worker count decides the whole in-processing batches and the in-processing batches account for the memory usage.

As this commit touches the essential part of the transaction code, please review it **extremely** carefully and be patient.

### Related changes

https://github.com/pingcap/tidb/pull/15053

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @sticnarf @coocood

File: store/tikv/2pc.go [hunk 1]
```diff
@@ -226,6 +262,296 @@ func newTwoPhaseCommitter(txn *tikvTxn, connID uint64) (*twoPhaseCommitter, erro
...
+
+	// Take the first batch.
+	batch, noMore, err := c.takeBatchMutationFromIterator(bo, action, &bi)
+	if err != nil || batch.mutations.len() == 0 {
+		return err
+	}
+	// Set the primary key.
+	batch.isPrimary = true
+	c.once.Do(func() {
+		c.primaryKey = batch.mutations.keys[0]
```
Comment (coocood) [inline store/tikv/2pc.go]: The first key may not be the primary key.

Comment (tiancaiamao) [reply store/tikv/2pc.go]: Yes, so this PR only handle the optimistic transaction case, the pessimistic is not handled yet.

**Comment by tiancaiamao** [comment]
Replaced by https://github.com/pingcap/tidb/pull/19055

**Pull request closed without merging**